STAT540 Seminar 06: Fitting and interpreting linear models (high volume)
---
_Rebecca Johnston 12/02/2014_


* <dim id="1a">[Load required packages and photoRec data](#1b)
* <dim id="2a">[Fit a linear model](#2b)
* <dim id="3a">[Be the boss of topTable()](#3b)
* <dim id="4a">[Perform inference for some contrasts](#4b)
* <dim id="5a">[Take-home exercise](#5b)


### <dim id="1b">[Load required packages and photoRec data](#1b)

```{r warning = FALSE}
library(limma) # fit linear models
library(ggplot2) # graphing
library(limma) # graphing
library(reshape2) # for reshaping data from wide to tall format
library(plyr) # for data aggregation tasks
```

Load gene expression dataset:
```{r}
prDat <- read.table("GSE4051_data.tsv")
str(prDat, max.level = 0)
```

Load design matrix:
```{r}
prDes <- readRDS("GSE4051_design.rds")
str(prDes)
```


### <dim id="2b">[Fit a linear model](#2b)

Explain gene expression in the wild type mice as a function of developmental stage (one-way ANOVA)?

Let's just work with the wild type data.
```{r}
wtDat <- subset(prDat, select = prDes$gType == "wt")
str(wtDat, max.level = 0)
wtDes <- subset(prDes, gType == "wt")
str(wtDes)
```

Before we can use limma we must make our design matrix. Let's accept the default "reference + treatment effects" scheme for handling the devStage factor. I encourage you to inspect the design matrix and confirm it's what you expect.
```{r}
wtDesMat <- model.matrix(~ devStage, wtDes)
wtDesMat
```

Now we will fit the model, for all probes at once, and use eBayes() to moderate the estimated error variances:
```{r}
wtFit <- lmFit(wtDat, wtDesMat)
wtEbFit <- eBayes(wtFit)
```

The first thing we might ask is "which genes show differential expression over the course of development"? This can be addressed with an overall F test for the model. In the language used in lecture, we will compare a "big" model to a "small" model, where the "big" model includes a mean parameter (or effect) for each level of devStage and the "small" model includes a single mean parameter, e.g. an intercept. You might expect this to be the F test performed by topTable() by default, i.e. when no specific coefficients or contrasts are given to the coef argument ...
```{r}
topTable(wtEbFit)
```

You'll see that, by default, topTable() reports the top 10 hits. But let's take more care and specify explicitly the coefficients we want to test for equality with zero.
```{r}
colnames(coef(wtEbFit))  # remind yourself of the coef names
(dsHits <- topTable(wtEbFit, coef = grep("devStage", colnames(coef(wtEbFit)))))
```

You will notice that these are not the same hits we got with our first call to topTable(). Compare, e.g., the Affy IDs for the top hits and/or look at the typical F statistic magnitudes. And so we learn that you really must use the coef argument (or a contrasts workflow in more complicated settings) to explicitly define *what you regard as a hit*.

Use the hit list you stored above and your functions for extracting and plotting data to produce this plot for hits 3, 6, and 9 on the list.

Function for preparing gene expression data:
```{r tidy = FALSE}
prepData <- function(x) {
  luckyDf <- prDat[x, ]
  luckyDf <- data.frame(gene = rownames(luckyDf), luckyDf) 
  luckyDf <- melt(luckyDf, id.vars = "gene", variable.name = "sidChar", 
                  value.name = "gExp")
  # Merge with wt experimental design
  luckyDf <- merge(luckyDf, wtDes, by = "sidChar")
  luckyDf <- arrange(luckyDf, devStage, gType)
  }
```

Function for making strip plots:
```{r fig.align='center', fig.height=8, fig.width=8}
makeStripplot <- function(x) {
  ggplot(x, aes(devStage, gExp, colour = gType, group = gType)) + 
    geom_point() + 
    facet_wrap(~ gene) +
    stat_summary(fun.y = mean, geom = "line")
  }
```

Run both functions given required gene list:
```{r fig.align='center', fig.height=6, fig.width=8}
makeStripplot(newDat <- prepData(rownames(dsHits[c(3, 6, 9),])))
str(newDat)
```

Does it look plausible to you that -- using only wild type data -- these probes show the most compelling evidence for expression change over development?

Optional exercise: use lm() on one or all 3 of these probes and check if the F stats and p-values are similar. Don't expect exact equality because you must remember that limma has moderated the estimated error variance.

I'll choose one of the 3 probes, "1451617_at", for a comparison:
```{r}
mDat <- prepData("1451617_at")
str(mDat)
```

Now call linear model using "lm" function:
```{r}
mDatlm <- lm(gExp ~ devStage, mDat)
summary(mDatlm)
```

Compare this to limma result:
```{r}
dsHits["1451617_at", c("F", "P.Value")]
```

The results from lm vs. limma for this probe are fairly similar, but not exactly the same.


### <dim id="3b">[Be the boss of topTable()](#3b)

You need to learn to take control of topTable() by using various arguments to get the hits you want in the order you want. Furthermore, you should familiarize yourself with the output it returns, so you are comfortable extracting the output that you need.

How many probes have Benjamini-Hochberg ("BH") adjusted p-values for the F test conducted above that are less than 1e-05?
```{r}
cutoff <- 1e-05
dsHits <- topTable(wtEbFit,
                   coef = grep("devStage", colnames(coef(wtEbFit))),
                   p.value = cutoff, n = nrow(wtDat))
nrow(dsHits)
```

What is the 63rd hit on this list? Provide it's Affy ID, F statistic, BH adjusted p-value, and the estimated effect for developmental stage "P6" *in that order*.
```{r}
dsHits[63, c("F", "adj.P.Val", "devStageP6")]
```

Consider the effects associated with developmental stages P2 and P10. Scatterplot the t statistics for the test that the P2 effect is zero against that for P10. Ideally this plot would be a high-volume scatterplot, include an x = y line, and have an aspect ratio of 1 and common axes, but just do your best.

First identify differentially expressed genes in P2 and P10 using `topTable`, make a dataframe containing the `t` statistic columns from the respective results, then plot using `ggplot2`:
```{r tidy = FALSE}
# colnames(wtEbFit)
P2Hits <- topTable(wtEbFit, coef = "devStageP2", n = nrow(wtDat), 
                   sort = "none")
P10Hits <- topTable(wtEbFit, coef = "devStageP10", n = nrow(wtDat), 
                    sort = "none")
# head(P2Hits)
# head(P10Hits)

P2P10t <- data.frame(P2 = P2Hits$t, P10 = P10Hits$t)
summary(P2P10t)
```

P10 has the largest range in values, therefore use these values for both the x and y limits of the plot for the aspect ratio to be 1:
```{r tidy=FALSE, fig.align='center', fig.height=10, fig.width=10}
ggplot(P2P10t, aes(P2, P10)) +
  geom_point(alpha = 0.1) +
  stat_density2d(geom = "polygon", aes(fill = ..level..)) + 
  geom_abline(intercept = 0, slope = 1) +
  scale_x_continuous(limits = c(min(P2P10t$P10), max(P2P10t$P10))) +
  scale_y_continuous(limits = c(min(P2P10t$P10), max(P2P10t$P10))) +
  xlab("t-statistic for P2 effect") +
  ylab("t-statistic for P10 effect")
```

Create a density plot of the associated adjusted p-values, so you can get a sense of which developmental stage, P2 or P10, is more clearly distinguished from baseline E16.
```{r message = FALSE, tidy = FALSE}
P2P10adjP <- data.frame(P2 = P2Hits$adj.P.Val, P10 = P10Hits$adj.P.Val)
P2P10adjPtall <- melt(P2P10adjP, variable.name = "DevStage", 
                      value.name = "adjPVal")
```

```{r fig.align='center', fig.height=8, fig.width=8}
ggplot(P2P10adjPtall, aes(adjPVal, colour = DevStage)) +
  geom_density() +
  xlab("Adjusted P Value")
```

Is this what you'd expect? Yes, I would expect the gene expression of P10 to be more clearly distinguished from baseline E16 than P2 as it is a later time point.

If you require a BH adjusted p-value less than 1e-03, how many hits do you get for P2? How many for P10? How much overlap is there?
```{r}
addmargins(table(P2Hits$adj.P.Val < 1e-03, P10Hits$adj.P.Val < 1e-03, 
                 dnn = c("P2", "P10")))
```

There are 53 hits for P2, 747 hits for P10, and 53 overlap.
> How can you do this test without using `table`?

Now just focus on the P10 effect. Create a scatterplot matrix of raw p-values, BH adjusted p-values, and BY p-values.
```{r tidy = FALSE}
P10pVals <- data.frame(raw = P10Hits$P.Value,
                       BH = P10Hits$adj.P.Val,
                       BY = p.adjust(P10Hits$P.Value, method = "BY"))
head(P10pVals)
```

The ggplot2 result from `plotmatrix` looked awful, so I'm resorting to base graphics. A message from the ggplot2 output suggested the `GGally` package intsead but this package masks several functions I use regularly such as `melt`!
```{r fig.align='center'}
pairs(P10pVals)
```

As expected, the raw and BH-adjusted p values are positively correlated. I'm not sure what BY-adjusted P values are so I can't comment on this.


### <dim id="4b">[Perform inference for some contrasts](#4b)

Let's try to distinguish genes that have stable expression at the last three developmental stages (P6, P10, and 4\_weeks) from those that do not. If expression doesn't change from P6 to P10 to 4\_weeks, then the effects for all 3 of those developmental stages should be the same. That means that the difference between the P10 and P6 effects is zero and ditto for the difference between 4_weeks effect and P10 (or P6, for that matter). Let's form these contrasts.

```{r tidy = FALSE}
colnames(wtDesMat)
(cont.matrix <- makeContrasts(
  P10VsP6 = devStageP10 - devStageP6, 
  fourweeksVsP10 = devStage4_weeks - devStageP10, 
  levels = wtDesMat))
wtFitCont <- contrasts.fit(wtFit, cont.matrix)
wtEbFitCont <- eBayes(wtFitCont)
```

What does `topTable()` do with our contrasts?
```{r}
topTable(wtEbFitCont)
```

The top hits are probes where there is big change from P6 to P10, from P10 to 4_weeks, or both. Let's check that by plotting the data from the top 4 hits.
```{r fig.align='center', fig.height=8, fig.width=8}
contHits <- topTable(wtEbFitCont)
makeStripplot(prepData(rownames(contHits)[1:4]))
```

So far, so good. These 4 probes show little expression change from P6 to P10 and a strong increase from P10 to 4_weeks. I would like to find some where there's a change in each case but perhaps in opposite direction. Let's press on.

Let's use decideTests() to adjust the p-values for both contrasts globally, i.e. all together and then threshhold them at a cutoff of 1e-04.
```{r}
cutoff <- 1e-04
wtResCont <- decideTests(wtEbFitCont, p.value = cutoff, method = "global")
summary(wtResCont)
```

We see there are 4 probes that go down from P6 to P10 and no hits going the other way. There are 8 probes that go down from P10 to 4_weeks and 46 going the other way. Let's try to pull out various hits and plot their data. Here are the 4 that decline from P6 to P10:
```{r fig.align='center', fig.height=8, fig.width=8}
(hits1 <- rownames(prDat)[which(wtResCont[, "P10VsP6"] < 0)])
makeStripplot(prepData(hits1))
```

Here are 4 of the 8 that decline from P10 to 4_weeks:
```{r fig.align='center', fig.height=8, fig.width=8}
(hits2 <- rownames(prDat)[which(wtResCont[, "fourweeksVsP10"] < 0)])
makeStripplot(prepData(hits2[1:4]))
```

Is there any overlap between these probes?
```{r}
intersect(hits1, hits2)
```

Apparently not. Here are 4 of the 46 that increase from P10 to 4_weeks:
```{r fig.align='center', fig.height=8, fig.width=8}
(hits3 <- rownames(prDat)[which(wtResCont[, "fourweeksVsP10"] > 0)])
makeStripplot(prepData(hits3[1:4]))
```

Is there any overlap between these probes and the previous "down" hits?
```{r}
intersect(hits1, hits3)
intersect(hits2, hits3)
```

That's disappointing. If I revisit this workflow but make the p-value cutoff less stringent, maybe I can find the gene expression profile I'm looking for:
```{r fig.align='center', fig.height=8, fig.width=8}
cutoff <- 0.01
nHits <- 8
wtResCont <- decideTests(wtEbFitCont, p.value = cutoff, method = "global")
summary(wtResCont)
hits1 <- rownames(prDat)[which(wtResCont[, "P10VsP6"] < 0)]
makeStripplot(prepData(hits1[1:nHits]))
hits2 <- rownames(prDat)[which(wtResCont[, "fourweeksVsP10"] < 0)]
makeStripplot(prepData(hits2[1:nHits]))
hits3 <- rownames(prDat)[which(wtResCont[, "P10VsP6"] > 0)]
makeStripplot(prepData(hits3[1:nHits]))
hits4 <- rownames(prDat)[which(wtResCont[, "fourweeksVsP10"] > 0)]
makeStripplot(prepData(hits4[1:nHits]))
vennDiagram(wtResCont)
```

```{r fig.align='center', fig.height=9, fig.width=10}
hits5 <- rownames(prDat)[which(wtResCont[, "P10VsP6"] != 0 &
                                  wtResCont[, "fourweeksVsP10"] != 0)]
makeStripplot(prepData(hits5))
```

```{r fig.align='center'}
hits6 <- rownames(prDat)[which(wtResCont[, "P10VsP6"] > 0 &
                                  wtResCont[, "fourweeksVsP10"] < 0)]
makeStripplot(prepData(hits6))
```


### <dim id="5b">[Take-home exercise](#5b)
See if you can find one or more probes that have some expression changes up to P6 and then hold steady all the way to 4_weeks.

Let's make a new contrast matrix and make all contrasts across the developmental stages:
```{r tidy = FALSE}
(cont.matrix <- makeContrasts(
  P2VsE16 = devStageP2 - Intercept, 
  P6VsP2 = devStageP6 - devStageP2, 
  P10VsP6 = devStageP10 - devStageP6, 
  fourweeksVsP10 = devStage4_weeks - devStageP10,
  fourweeksVsP6 = devStage4_weeks - devStageP6,
  levels = wtDesMat))
wtFitCont <- contrasts.fit(wtFit, cont.matrix)
wtEbFitCont <- eBayes(wtFitCont)
cutoff <- 1e-03
wtResCont <- decideTests(wtEbFitCont, p.value = cutoff, method = "global")
summary(wtResCont)
# head(wtResCont)
```

Find hits where contrasts between P2 vs E16 and P6 vs P2 are not equal to 0:
```{r fig.align='center', fig.height=9, fig.width=10}
hits7 <- rownames(prDat)[which(
  wtResCont[, "P2VsE16"] < 0 &
    wtResCont[, "P6VsP2"] < 0 & 
    wtResCont[, "P10VsP6"] == 0 & 
    wtResCont[, "fourweeksVsP10"] == 0 &
    wtResCont[, "fourweeksVsP6"] == 0)]
makeStripplot(prepData(hits7))
```
